### PR TITLE
Update apps generator plugin dependencies

### LINKED ISSUE #498: Update dependencies in the apps metadata plugin
Some of the dependencies in the Apps metadata and docs generator plugin used by the stream-applications need to be updated. 

One example, is where our good friend dependa-bot recently created these to update Maven:

- https://github.com/spring-cloud/stream-applications/pull/491
- https://github.com/spring-cloud/stream-applications/pull/490

After the 4.x release, let's update the plugins dependencies across the board.

### PR DESCRIPTION
**Pull request by onobc**
This commit updates all library and plugin dependencies in the `spring-cloud-dataflow-apps-plugin` modules to the latest versions.

Additionally:
- performs a general cleanup of build and README files.
- updates to maven 3.9.9

Resolves #498

### CODE REVIEW

File: spring-cloud-dataflow-apps-plugin/.mvn/wrapper/maven-wrapper.properties [hunk 1]
```diff
@@ -1,2 +1,20 @@
-distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.6.3/apache-maven-3.6.3-bin.zip
-wrapperUrl=https://repo.maven.apache.org/maven2/io/takari/maven-wrapper/0.5.6/maven-wrapper-0.5.6.jar
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (onobc) [inline spring-cloud-dataflow-apps-plugin/.mvn/wrapper/maven-wrapper.properties:1]: Updated the wrapper and maven to 3.9.9

File: spring-cloud-dataflow-apps-plugin/pom.xml [hunk 2]
```diff
@@ -3,41 +3,44 @@
 	<modelVersion>4.0.0</modelVersion>
 	<groupId>org.springframework.cloud</groupId>
 	<artifactId>spring-cloud-dataflow-apps-plugin-parent</artifactId>
-	<version>1.0.15-SNAPSHOT</version>
+	<version>1.1.0-SNAPSHOT</version>
```
Comment (onobc) [inline spring-cloud-dataflow-apps-plugin/pom.xml:6]: Bumping up minor version to indicate the dependency updates

File: spring-cloud-dataflow-apps-plugin/spring-cloud-dataflow-apps-docs-plugin/src/main/java/org/springframework/cloud/dataflow/app/plugin/ConfigurationMetadataDocumentationMojo.java [hunk 3]
```diff
@@ -418,14 +418,14 @@ public Manifest getManifest() {
 		}
 
 		@Override
-		public List<Archive> getNestedArchives(EntryFilter ignored) throws IOException {
+		public Iterator<Archive> getNestedArchives(EntryFilter ignored, EntryFilter ignored2) throws IOException {
```
Comment (onobc) [inline spring-cloud-dataflow-apps-plugin/spring-cloud-dataflow-apps-docs-plugin/src/main/java/org/springframework/cloud/dataflow/app/plugin/ConfigurationMetadataDocumentationMojo.java:421]: API change in Maven

File: spring-cloud-dataflow-apps-plugin/spring-cloud-dataflow-apps-generator-plugin/src/main/java/org/springframework/cloud/dataflow/app/plugin/MavenXmlWriter.java [hunk 4]
```diff
@@ -103,32 +103,37 @@ public static String toXml(Plugin plugin) {
...
 			serializer.setProperty("http://xmlpull.org/v1/doc/properties.html#serializer-line-separator", "\n");
 			serializer.setOutput(writer);
-
 			serializer.startDocument("UTF-8", null);
 			elementWriter.accept(serializer);
 			serializer.endDocument();
-
 			String result = writer.toString();
-			return result.substring(result.indexOf('\n') + 1); // remove first line (e.g. remove the <?xml ... ?>)
+			// Strip the namespace info
```
Comment (onobc) [inline spring-cloud-dataflow-apps-plugin/spring-cloud-dataflow-apps-generator-plugin/src/main/java/org/springframework/cloud/dataflow/app/plugin/MavenXmlWriter.java:129]: Latest version of `MXSerializer` handles generated xml a bit different. This impl has been improved to look for the actual tag, not the 1st newline character for parsing.

File: spring-cloud-dataflow-apps-plugin/spring-cloud-dataflow-apps-generator-plugin/src/main/java/org/springframework/cloud/dataflow/app/plugin/MavenXmlWriter.java [hunk 5]
```diff
@@ -168,6 +173,9 @@ public static <T> void invokeMavenXppWriteMethod(T modelElementToWrite, String w
 	public static String indent(String input, int indentation) {
 		String indentPrefix = "\n" + IntStream.range(0, indentation).mapToObj(i -> " ").collect(Collectors.joining());
 		String indentedInput = input.replace("\n", indentPrefix);
-		return indentedInput.substring(0, indentedInput.lastIndexOf(indentPrefix)); // remove the last empty line.
+		if (input.endsWith(indentPrefix)) {
```
Comment (onobc) [inline spring-cloud-dataflow-apps-plugin/spring-cloud-dataflow-apps-generator-plugin/src/main/java/org/springframework/cloud/dataflow/app/plugin/MavenXmlWriter.java:176]: Similar to comment above, more lenient processing.

File: spring-cloud-dataflow-apps-plugin/spring-cloud-dataflow-apps-generator-plugin/src/test/java/org/springframework/cloud/dataflow/app/plugin/MojoHarnessTest.java [hunk 6]
```diff
@@ -25,78 +25,72 @@
...
-
-		File pomRoot = new File("src/test/resources/unit/http-source-apps/");
-
-		SpringCloudStreamAppGeneratorMojo myMojo = (SpringCloudStreamAppGeneratorMojo)
-				mojoRule.lookupConfiguredMojo(pomRoot, "generate-app");
-
+	void httpSourceAppsGeneration() throws Exception {
+		File pomRoot = new File("src/test/resources/unit/http-source-apps");
+		MavenProject project = readMavenProject(pomRoot);
+		SpringCloudStreamAppGeneratorMojo myMojo = (SpringCloudStreamAppGeneratorMojo) super.lookupConfiguredMojo(project, "generate-app");
```
Comment (onobc) [inline spring-cloud-dataflow-apps-plugin/spring-cloud-dataflow-apps-generator-plugin/src/test/java/org/springframework/cloud/dataflow/app/plugin/MojoHarnessTest.java:60]: There is no Junit5 replacement for the Mojo Junit5 rule. **HOWEVER**, the `AbstractMojoTestCase` has what we need but we have to jump through some hoops in `readMavenProject` to get an actual project object to then pass into `super.lookupConfiguredMojo`.

File: spring-cloud-dataflow-apps-plugin/spring-cloud-dataflow-apps-docs-plugin/pom.xml [hunk 7]
```diff
@@ -5,90 +5,66 @@
...
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-loader-classic</artifactId>
 		</dependency>
 		<dependency>
-			<groupId>org.assertj</groupId>
-			<artifactId>assertj-core</artifactId>
-			<version>3.15.0</version>
+			<groupId>org.springframework.cloud</groupId>
+			<artifactId>spring-cloud-dataflow-configuration-metadata</artifactId>
+			<version>2.11.5</version>
```
Comment (onobc) [inline spring-cloud-dataflow-apps-plugin/spring-cloud-dataflow-apps-docs-plugin/pom.xml:36]: This is way less than ideal. However, it was previously hardcoded to `2.7.0` so just updated to `2.11.5` for now. Need to figure out how to detangle this bidrectional dependency (apps plugin needs SCDF and SCDF needs apps plugin).

File: spring-cloud-dataflow-apps-plugin/spring-cloud-dataflow-apps-metadata-plugin/src/main/java/org/springframework/cloud/dataflow/app/plugin/MetadataAggregationMojo.java [hunk 8]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2017-2020 the original author or authors.
+ * Copyright 2017-2024 the original author or authors.
```
Comment (artembilan) [inline spring-cloud-dataflow-apps-plugin/spring-cloud-dataflow-apps-metadata-plugin/src/main/java/org/springframework/cloud/dataflow/app/plugin/MetadataAggregationMojo.java:2]: DITTO - bogus.

**Review by artembilan — CHANGES_REQUESTED** [review]
Cannot  tell that I understood too much from the change, but mostly it is really OK.
Just a couple nit-picks from me.

Thanks

File: spring-cloud-dataflow-apps-plugin/spring-cloud-dataflow-apps-generator-plugin/src/main/resources/template/AppTests.java [hunk 9]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2019-2020 the original author or authors.
+ * Copyright 2019-2024 the original author or authors.
```
[the remark this answers is not in the document]
Comment (onobc) [reply spring-cloud-dataflow-apps-plugin/spring-cloud-dataflow-apps-generator-plugin/src/main/resources/template/AppTests.java:2]: Typically I would not just update the copyright like this but I was basically brining this repo up-to-date for everything I could find that was outdated. I figured I would catch up the copyrights as well.

File: spring-cloud-dataflow-apps-plugin/spring-cloud-dataflow-apps-metadata-plugin/README.adoc [hunk 10]
```diff
@@ -1,84 +1,85 @@
...
+----
 <build>
   <plugins>
       <plugin>
         <groupId>org.springframework.cloud</groupId>
-        <artifactId>spring-cloud-app-starter-metadata-maven-plugin</artifactId>
-        <version>2.0.0.BUILD-SNAPSHOT</version>
-
+        <artifactId>spring-cloud-dataflow-apps-metadata-plugin</artifactId>
+        <version>1.0.0</version>
```
[the remark this answers is not in the document]
Comment (onobc) [reply spring-cloud-dataflow-apps-plugin/spring-cloud-dataflow-apps-metadata-plugin/README.adoc:70]: Good catch. Will update

File: spring-cloud-dataflow-apps-plugin/spring-cloud-dataflow-apps-metadata-plugin/src/main/java/org/springframework/cloud/dataflow/app/plugin/MetadataAggregationMojo.java (hunk 8 above)
Comment (onobc) [reply spring-cloud-dataflow-apps-plugin/spring-cloud-dataflow-apps-metadata-plugin/src/main/java/org/springframework/cloud/dataflow/app/plugin/MetadataAggregationMojo.java:2]: See https://github.com/spring-cloud/stream-applications/pull/593#discussion_r1878558907

**Comment by artembilan** [comment]
I have no idea what is that CI build failure:
```
[Info] Cleaning build info...
[Info] Cleaned build info spring-cloud-dataflow-apps-plugin-main-snapshot/22.
[Info] Running Mvn...
[Info] The build-info-extractor jar is not cached locally. Downloading it now...
 You can set the repository from which this jar is downloaded. Read more about it at https://www.jfrog.com/confluence/display/CLI/CLI+for+JFrog+Artifactory#CLIforJFrogArtifactory-DownloadingtheMavenandGradleExtractorJARs
[Info] Downloading build-info-extractor from https://oss.jfrog.org/artifactory/oss-release-local/org/jfrog/buildinfo/build-info-extractor-maven3/2.25.0/build-info-extractor-maven3-2.25.0-uber.jar
Error:  Server ID 'repo.spring.io' does not exist.
Error: Process completed with exit code 1.
```

**Comment by onobc** [comment]
> I have no idea what is that CI build failure:
> 
> ```
> [Info] Cleaning build info...
> [Info] Cleaned build info spring-cloud-dataflow-apps-plugin-main-snapshot/22.
> [Info] Running Mvn...
> [Info] The build-info-extractor jar is not cached locally. Downloading it now...
>  You can set the repository from which this jar is downloaded. Read more about it at https://www.jfrog.com/confluence/display/CLI/CLI+for+JFrog+Artifactory#CLIforJFrogArtifactory-DownloadingtheMavenandGradleExtractorJARs
> [Info] Downloading build-info-extractor from https://oss.jfrog.org/artifactory/oss-release-local/org/jfrog/buildinfo/build-info-extractor-maven3/2.25.0/build-info-extractor-maven3-2.25.0-uber.jar
> Error:  Server ID 'repo.spring.io' does not exist.
> Error: Process completed with exit code 1.
> ```

Digging in..

**Comment by onobc** [comment]
@artembilan this https://github.com/spring-cloud/stream-applications/commit/5cce183e1ba25335193bb057fb37301b05b9f3a1#diff-56d907718f7f9fbb26e09bddd3723c4ba7782bf2eaf3b8f22781f2a66df9701bR43 fixed it

Something has changed on Artifactory where the server id used to generate the token was "repo.spring.io" but is now different per group of projects (e.g. it is now `github-spring-cloud` for most Spring Cloud projects).

**Pull request merged**